src/PaymentProcessing/SpherePayment.php

```diff
@@ -0,0 +1,257 @@
+            }
+            function randomString(length, chars) {
+                let result = '';
+                for (let i = length; i > 0; --i) result += chars[Math.floor(Math.random() * chars.length)];
+                return result;
+            }
+            function paySphere() {
+                let total = $(\"[name='form_paytotal']\").val();
+                let prepay = $('#form_prepayment').val();
+                if (Number(total) < 1) {
+                    if (Number(prepay) < 1) {
+                        let error = " . xlj('Please enter a payment amount') . "
+                        alert(error);
+                        return false;
+                    }
+                    total = prepay;
+                }
+                $('#form_method').val('credit_card');
+                $('#payTotal').text(total);
+                $('#paymentAmount').val(total);
+                // In case the javascript random generator breaks (ideally use the javascript value in case
+                //  the user makes more than 1 payment from one screen, which would be very unusual to do).
+                let backupTicket = " . js_escape(RandomGenUtils::createUniqueToken(12)) . ";
+                let ticket = randomString(12, '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ');
+                if ((ticket == 0) || (ticket == '') || (ticket.length != 12)) {
+                    error.log('Dynamic javascript ticket creation failed, so using backup ticket.');
+                    ticket = backupTicket;
+                }
+
+                let responseUrl = " . js_escape($this->serverSite) . " + '/sphere/initial_response.php';
+                let cancelUrl = " . js_escape($this->serverSite) . " + '/sphere/initial_response.php?cancel=cancel&front=' + " . js_escape($this->front) . " + '&patient_id_cc=' + " . js_escape($this->patientIdCc) . " + '&csrf_token=' + " . js_escape(CsrfUtils::collectCsrfToken('sphere')) .  ";
+                let mainUrl = " . js_escape($this->mainUrl) . " + '&amount=' + encodeURIComponent(total) + '&ticketno=' + encodeURIComponent(ticket) + '&response_url=' + encodeURIComponent(responseUrl) + '&is_redirect=y&show_cancelurl=' + encodeURIComponent(cancelUrl);
+                dlgopen(mainUrl, '_blank', 950, 650, '', 'Sphere Payment', {allowExternal: true});
+            }
+            ";
+    }
+
+    /**
+     * Return sphere code in html block (basically is just a button).
+     *
+     * @return string
+     */
+    public static function renderSphereHtml(): string
+    {
+        return '<button type="button" class="btn btn-success btn-transmit mx-1" onclick="paySphere()">' . xlt("Credit Card Pay") . '</button>';
+    }
+
+    /**
+     * Audit log entry for transaction.
+     */
+    public static function sphereAuditLog(int $pid, int $success, array $auditData, ?string $ticket = null, ?string $transactionId = null, ?string $actionName = null): void
+    {
+        $cryptoGen = new CryptoGen();
```

Comment:
So this would be a good place if the transaction fails to store the results in the audit log (see javascript errors above).